File: Java/986. Interval List Intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
        List<int[]> ans = new ArrayList<>();
        int p1 = 0, p2 = 0;
        
        while(p1 < firstList.length && p2 < secondList.length) {
            int left = Math.max(firstList[p1][0], secondList[p2][0]);
            int right = Math.min(firstList[p1][1], secondList[p2][1]);
            
            if(left <= right) {
                ans.add(new int[] {left, right});
            }
            
            if(firstList[p1][1] < secondList[p2][1]) {
                p1++;
            } else {
                p2++;
            }
        }
        
        int[][] res = new int[ans.size()][];
        int i = 0;
        
        for(int[] a: ans) {
            res[i++] = a;
        }
        
        return res;
    }
}

File: Python3/986. Interval List Intersections.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        p1 = p2 = 0
        ans = []
        
        while p1 < len(firstList) and p2 < len(secondList):
            left = max(firstList[p1][0], secondList[p2][0])
            right = min(firstList[p1][1], secondList[p2][1])
            
            if left <= right:
                ans.append([left, right])
            
            # keep the later one
            if firstList[p1][1] < secondList[p2][1]:
                p1 += 1
            else:
                p2 += 1
                    
        return ans